docs/adr/0014-auth-for-settings-redesign.md

```diff
@@ -0,0 +1,96 @@
+# Authentication in Settings Redesign
+
+- Status: accepted
+- Deciders: Les Orchard, Lauren Zugai, Ben Bangert, Ryan Kelly
+- Date: 2020-04-15
+
+Issue: https://jira.mozilla.com/browse/FXA-1559
+
+## Context and Problem Statement
+
+The Settings Redesign app needs to query & mutate the same protected user data as the existing settings app hosted on content-server. This will require some form of authentication & authorization to manage that data.
+
+## Decision Drivers
+
+- Smooth UX
+- Security
+- Development velocity
+- Ease of integration
+
+## Considered Options
+
+- Authenticate via OAuth2 to use auth-server APIs
+- Authenticate via OAuth2 to use DB directly
+- Reuse existing sessionToken on content-server with auth-server APIs
+- Implement own login sessions
+
+## Decision Outcome
+
+Chosen option: "Authenticate via OAuth2 to use auth-server APIs", because:
+
+- It requires minimal changes to existing auth-server implementation and infrastructure.
+- It relies on an authentication mechanism with relatively well-known security properties.
+
+### Positive Consequences
+
+- The OAuth2 access token mechanism is better suited to this purpose than a novel scheme to share session token credentials.
+- The new Settings Redesign app can use existing auth-server APIs with minimal modifications to accept scoped OAuth2 access tokens. This can constrain most of the novelty in the project to the redesigned & reimplemented settings UX.
+
+### Negative Consequences
+
+- We will need to modify auth strategies in auth-server APIs, taking care not to affect existing usage in the production settings app.
+- We don't have an entirely greenfield project, which could be a bit of a drag. But, rediscovering lessons learned in existing code can also be a drag.
+
+## Pros and Cons of the Options
+
+### Authenticate via OAuth2 to use auth-server APIs
+
+We would like to someday deprecate the session token mechanism altogether. This could be a start.
+
+- Pros:
+  - Security properties of OAuth2 authn / authz are known and not entirely novel
+  - Auth-server APIs can be reused with small changes
+- Neutral:
+  - GraphQL server needs to proxy auth-server APIs
+- Cons:
+  - APIs on auth-server need auth strategy changes to accept OAuth tokens
+  - May initially require a hack similar to payments-server where a pre-generated access token is passed along
```

Comment:
Is this still the case, or will we use the sessionToken from localStorage to then generate the access token?

And just to clarify, if/when we do implement `prompt=none` for internal RPs like payments/settings, would we simply not need to pass along that content-server generated OAuth to payments, and then not need to generate an access token for settings? It does sound like that's worth discussing as this will be the second app needing some slight hackery without it (or at least a less ideal solution), but I'm not sure what the scope of remaining work there is. Depending on that scope, maybe that could be an option here... If we're granted the time by management, maybe that's the path to head down? Or, maybe this is the best direction for now and `prompt=none` discussions can be a follow up since the scope of the Settings Redesign project is already a beast in itself.

Reply:
> Is this still the case, or will we use the sessionToken from localStorage to then generate the access token?

Well, "use the sessionToken ... to then generate the access token" **is** the hack we're using within content-server 😅  That's not the normal flow for OAuth.

> And just to clarify, if/when we do implement prompt=none for internal RPs like payments/settings, would we simply not need to pass along that content-server generated OAuth to payments, and then not need to generate an access token for settings?

We'd still need an access token for settings, it's just that we'd acquire it using the normal OAuth flow with a `?prompt=none` - related to stuff @6a68 has [been working on](https://github.com/mozilla/fxa/pull/4818).

> Or, maybe this is the best direction for now and prompt=none discussions can be a follow up since the scope of the Settings Redesign project is already a beast in itself.

Yeah, this is what I'd suggest doing. We already use the pre-generated access token hack in payments-server. We start with that on settings redesign and upgrade to prompt=none - both approaches result in an OAuth access token, so moving from one to the other shouldn't significantly change anything besides the top-level entry point to the app

Reply:
> It does sound like that's worth discussing as this will be the second app needing some slight hackery without it (or at least a less ideal solution), but I'm not sure what the scope of remaining work there is.

Oh but yeah, re: scope here - I'm not entirely sure either. I have a vague grasp of what proper use of prompt=none looks like. I had a better grasp a few months ago, but I think I flushed those parts of my brain out and replaced them with turnip prices from Animal Crossing. So we'd need a bit of refresh & review to see where we're at now.

From what I remember, there's some ID verification work required to use prompt=none that was impractical for payments-server purposes in accepting users coming from static marketing pages rather than RPs. But, I think it might be more practical for our new settings redesign RP to do it.

Reply:
I thought this would live under the existing domain so we wouldn't need to do an oauth dance here?

Reply:
> I thought this would live under the existing domain so we wouldn't need to do an oauth dance here?

Well, I think there are a few things we can do here:

1. Have content-server generate an access token for settings redesign [like it does for payments-server](https://github.com/mozilla/fxa/blob/master/packages/fxa-content-server/app/scripts/lib/payment-server.js#L38)
    - probably the quickest, since we're already using it for payments-server

1. Have settings redesign request the access token itself, using the FxA client and session token [just like content-server does](https://github.com/mozilla/fxa/blob/master/packages/fxa-content-server/app/scripts/models/account.js#L231)
    - doable, but pulls dependencies into settings redesign frontend for FxA client and Hawk authentication

1. Do an OAuth dance as an RP to get an access token
    - seems to me like where we want to end up eventually if we want move everything to OAuth and deprecate the session token scheme.

Reply:
I'm on board with this change if rfk sees a clear path forward, though I think we need the flow spec'd out in greater detail to decide whether it will work or not. The prompt=none code currently relies on a session token to decide if we should allow a request to proceed, but eliminating that token seems to be in the discussion here, and I'm a bit confused about how the pieces will all fit together.

Reply:
Although if they are on the same domain, I guess we can use an id token already stored client-side to supply to prompt=none as the id_token_hint to transparently log the user into the best settings site. But that assumes the session token stays in place.

Reply:
I don't think eliminating session token is in the immediate scope for settings redesign stuff - more of a future wish list item. That would deserve its own ADR and associated project. 

FWIW, I'd guess we'd never *entirely* eliminate a session token, because we need something to track the session of a user login and flow through OAuth dances. I think it's that we'd remove support for using the session token in nearly all of the auth-server APIs for everything else

Reply:
> Have content-server generate an access token for settings redesign like it does
> for payments-server

This seems OK to me, although it would be good to see the steps spelled out in more detail. Some things to consider:

* Since you're on the same domain, I would strongly consider passing this token via e.g. localStorage or some other local mechanism, rather than in the URL where it can be injected by external links.
* How will the new settings page refresh its OAuth token if it discovers that that the token has expired or has been revoked?
* What scope(s) will you use for this token? How will you handle the different levels of authentication that are currently possible on a sessionToken? (e.g. verified vs unverified sessions).

Reply:
> I'm not a big fan of the second option (i.e. have settings redesign frontend request the token) because it would pull in those dependencies and then only use them for that one thing.

Ah, OK, this answers one of my earlier questions not asked in the right way. I wasn't sure if the comment in the ADR, "May initially require a hack similar to payments-server where a pre-generated access token is passed along" was still the approach we were discussing if we could just directly request the token.

Sounds like we're leaning towards passing a pre-generated token to settings similar to payments but through localStorage instead? Curious about answers to Ryan's questions too.

Reply:
I was trying to be a little vague about the specific mechanism in this ADR, but maybe it would help to explicitly lay it out?

> I would strongly consider passing this token via e.g. localStorage

:+1: localStorage should work

> How will the new settings page refresh its OAuth token if it discovers that that the token has expired or has been revoked?

That's one of the problems on payments-server right now: We don't refresh it, we just give it a long expiration that hopefully covers the majority of uses. That's part of why we eventually want to resolve #1695.

For settings redesign, though, we could probably just invent our own quiet refresh API that generates a new token and stashes it in localStorage. Then, make managing that part of whatever general API abstraction we use in settings redesign

> What scope(s) will you use for this token? How will you handle the different levels of authentication that are currently possible on a sessionToken? (e.g. verified vs unverified sessions).

Do we have these different levels documented somewhere? (I don't know them all myself) I'd imagine we should gin up a scope for each of them? Probably also eventually need the settings redesign app to request tokens with escalating scopes as needed?

Reply:
> Do we have these different levels documented somewhere?

Not in a deliberate fashion that I know of, because they've grown organically over time :-/

I tried to include some of these details in WIP docs patch [here](https://github.com/mozilla/ecosystem-platform/pull/39/files) but haven't revisited that in a while.

In practice, some of the routes in auth-server have an unfortunate split where some auth decisions are made in the auth strategy hander (e.g. by requiring the "sessionToken" strategy) and some auth decisions are more inline in the route handler code (e.g. by checking whether the token is verified or not).

Here's an example of how things could go wrong, in the [`/totp/create` route](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/lib/routes/totp.js#L32). It currently uses the [`sessionToken` auth stragey](https://github.com/mozilla/fxa/blob/70333b9e5492f0ae41858eb65d96c71eadda85ff/packages/fxa-auth-server/lib/routes/totp.js#L35) but has an extra inline check the make sure that the sessionToken has been verified via email confirmation loop:

  https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/lib/routes/totp.js#L61-L63

If we naively add an second auth strategy to this endpoint so that it can be accessed via an OAuth token, then we would end up with a `request.auth.credentials` object that does not have a `tokenVerificationId` property, and would silently bypass that check! So we'll need to capture that same information in the OAuth token somehow, probably via scopes. We don't have to design all of the details of that today.

In practice, I expect we'll need to grep the auth-server code for places that throw the following errors, which indicate missing levels of authentication:

* `ACCOUNT_UNVERIFIED` / `error.unverifiedAccount()`
* `SESSION_UNVERIFIED` / `error.unverifiedSession()`
* `TOTP_REQUIRED` / `error.totpRequired()`

In places where there is inline route handler code that throws such errors, is probably a good opportunity to refactor that logic out into an auth strategy, to make the auth requirements more obvious to future readers.

@vbudhram can you think of any other cases besides these three?